### DISCUSSION TITLE
Toolkit: Support for process.env for build time env settings

### DISCUSSION THREAD

**Discussion by Simonl9l** [grafana/grafana#38179] [Dashboards, Panels & Visualizations]
Per [this](https://community.grafana.com/t/plugin-toolkit-support-for-dotenv/38867) - very frustrating that no one _**ever**_ (specifically from Grafana) replies on the community forums!

It would be great to add [dotenv](https://github.com/motdotla/dotenv) support to the plugin toolkit.

As developer we need to be able to ensure our plugins can easily target different environments in our product development workflow. We might, for example, have different URL to connect to for each of our Development, Testing or Production environments or other settings that are environment specific.

in the `package.json` it would be great to be able to specify for example

`"build:development": "grafana-toolkit plugin:build -mode development"`

this would pick up our `.env.development` (or `.env.developement.local`) configuration file and in our code we can then say use `process.env.MY_SETTING`

This is similar to what has been set up in the [Vue cli](https://cli.vuejs.org/guide/mode-and-env.html#modes)

**Comment by torkelo** [comment]
Sorry for not responding on the community site, we must have missed this. For feature requests like this gitub is better. 

We have not had any similar request that I am familiar with, normally different environment settings are stored in data source instance settings in Grafana instead (or in Grafana App settings). Not in build settings.

**Comment by Simonl9l** [comment]
Hi @torkelo Thanks for the speedy reply.

I'm a Grafana newbie...still to some degree getting my head around the toolkit. From a plugin develop experience perspective its a learning curve... 

We have a PasS platfrom that we're developing a data source to allow our customer to display data from that service, as one of many integrations.

We will need to "bake" our public endpoint into the release version (will submit with a pull request in time) as part of that build process. It's **_not_** a user configurable setting.

Our customers will use the _Data Source_ settings to define specific identifiers and keys to access the data in their Tenant in our system  via REST calls to that endpoint, and _Query Settings_ for their defined Data Sets input platform, as picked up by REST call to that endpoint.

The plugin will then pull observable streamed data from the data source to display in Grafana Dashboard our customer define.

So we need a way to make this endpoint (and other settings) configurable per our DevOps build environment (Development/Staging/Production) so we can take it though to testing cycles in each environment. 

As highlighted Vue make this very easy with `env.process` as do. many others.

Can you suggest an alternate approach? 

We'd hope that developer toolkits consider that scenario as Vue does - surely are we're not the only ones developing data sources for such a platform?

The toolkit is after all about eliminating developer friction that has the value proportion to Grafana to making it more sticky, as well complete API documentation, and more code examples, let alone a responsive and maintained community forum. 

Perhaps close the Community Develop forum and send everyone here...albeit with over 2.2k open issues and 111 open pull requests there is perhaps a bigger issue here.

Thoughts?

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by GuYounes** [comment]
Hi @Simonl9l,

You can create an new webpack.config.js, import their webpackconfig and override it this way.

`const grafanaWebpackConfigBuilder = require('@grafana/toolkit/src/config/webpack.plugin.config.js');`

```
return new Promise((resolve) => {
  grafanaWebpackConfigBuilder
    .loadWebpackConfig({
      watch: false,
      preserveConsole: true,
      production: isProd,
    })
    .then((config) => {
      // modify the config here...
      resolve(config)
    })
});
```

By doing that you can add plugins webpack plugins to handle your .env file.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
